In [3]:
# Imports and pip installations (if needed)
import pandas as pd

# Part 1: Load the dataset

In [4]:
# Load the given datasets

categorical = pd.read_csv("Data/chronic_kidney_disease_categorical.csv")
numerical = pd.read_csv("Data/chronic_kidney_disease_numerical.csv")
# Print the data

numerical
categorical

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,appet,pe,ane,Target
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,good,yes,no,ckd
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,good,no,no,notckd
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,good,yes,no,ckd
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,good,no,yes,ckd
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
425,769512,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,good,no,no,notckd
426,872034,2.0,1.0,abnormal,abnormal,notpresent,present,yes,yes,yes,poor,no,no,ckd
427,643007,NaN,NaN,NaN,normal,notpresent,notpresent,yes,yes,yes,poor,no,no,ckd
428,840392,NaN,NaN,NaN,NaN,notpresent,notpresent,yes,no,no,poor,no,no,ckd


# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The features are the columns of variables that help the model predict the label, which in this case is the target column. The features in this dataset are both categorical and numerical. 
Categorical columns:
1) sg : specific gravity
2) al : albumin
3) su : sugar
Numerical columns:
1) age
2) bp : blood pressure
3) bgr : blood glucose random

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [5]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows
numDuplicates = numerical[numerical.duplicated(keep=False)]
numDuplicates
# Drop these duplicate rows
numUnique = numerical.drop_duplicates()
# Repeat the same for categorical dataset. Print the duplicate rows and drop them
catDuplicates = categorical[categorical.duplicated(keep=False)]
catDuplicates
catUnique = categorical.drop_duplicates()

## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: There are unique ids present in both datasets. I believe this is the case because each id, representa a patient and then the data collected on that patient could be divided up into categorical and numerical data. However, I do not expect every unique_id to be included in both data sets since they have a different number of rows each after removing the duplicates. Thus I imagine the new table will have the same number of rows as the larger data set at this point. 

In [6]:
# Merge the two given numerical and categorical datasets based on their unique_ID.
combined = pd.merge(catUnique, numUnique, on='unique_id', how='inner')
#Print the combined dataset
combined

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
0,224481,3.0,NaN,NaN,NaN,notpresent,notpresent,yes,yes,no,...,60.0,125.0,21.0,1.3,137.0,3.4,15.0,46.0,NaN,NaN
1,992643,0.0,0.0,NaN,NaN,notpresent,notpresent,no,no,no,...,70.0,82.0,36.0,1.1,150.0,3.5,14.5,52.0,9400.0,6.1
2,308740,1.0,0.0,abnormal,abnormal,notpresent,notpresent,yes,no,no,...,70.0,139.0,29.0,1.0,NaN,NaN,10.5,32.0,NaN,NaN
3,450314,1.0,0.0,NaN,normal,notpresent,notpresent,no,no,no,...,NaN,NaN,16.0,0.7,138.0,3.2,8.1,NaN,NaN,NaN
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
385,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
386,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9
387,492993,3.0,0.0,NaN,normal,notpresent,notpresent,yes,no,no,...,80.0,76.0,162.0,9.6,141.0,4.9,7.6,24.0,3800.0,2.8
388,273038,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,117.0,22.0,1.2,138.0,3.5,13.0,45.0,5200.0,5.6


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [7]:
# Calculate the percentage of rows that contain atleast one missing value
incompleteRows = combined.isnull().any(axis=1)
percent = incompleteRows.mean() * 100
# Print %
print(f"\n{percent:.2f}%")
# Drop these rows from the dataset
cleanCombined = combined.dropna()
# Print the Dataset
cleanCombined


60.77%


,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
5,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
7,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7
8,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0
9,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,...,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
382,996076,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2
384,397388,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,...,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2
385,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
386,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9


## Part 3.4: Sort the dataset according to the Labels

In [8]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting
OGcleanCombined = cleanCombined
cleanCombined.sort_values(by='Target')
# Print the dataset
cleanCombined

,unique_id,al,su,rbc,pc,pcc,ba,htn,dm,cad,...,bp,bgr,bu,sc,sod,pot,hemo,pcv,wbcc,rbcc
4,881763,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,75.0,31.0,1.2,141.0,3.5,16.5,54.0,7800.0,6.2
5,114717,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,122.0,25.0,0.8,138.0,5.0,17.1,41.0,9100.0,5.2
7,901528,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,118.0,44.0,0.7,137.0,3.5,14.8,45.0,9300.0,4.7
8,433613,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,97.0,40.0,0.6,150.0,4.5,14.2,48.0,10500.0,5.0
9,621332,1.0,0.0,abnormal,normal,notpresent,notpresent,no,yes,no,...,80.0,303.0,35.0,1.3,122.0,3.5,10.4,35.0,10900.0,4.3
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
382,996076,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,60.0,105.0,49.0,1.2,150.0,4.7,15.7,44.0,10400.0,6.2
384,397388,2.0,2.0,normal,normal,notpresent,present,yes,no,yes,...,100.0,280.0,35.0,3.2,143.0,3.5,13.0,40.0,9800.0,4.2
385,789906,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,70.0,83.0,42.0,1.2,139.0,3.7,16.2,50.0,9300.0,5.4
386,490253,0.0,0.0,normal,normal,notpresent,notpresent,no,no,no,...,80.0,81.0,46.0,0.6,135.0,4.9,13.9,48.0,6900.0,4.9


## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [9]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
catDummies = pd.get_dummies(cleanCombined, columns=categorical_columns)
# Print the dataset #153x36
catDummies

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
4,881763,0.0,0.0,42.0,70.0,75.0,31.0,1.2,141.0,3.5,...,True,False,True,False,True,False,True,False,False,True
5,114717,0.0,0.0,41.0,80.0,122.0,25.0,0.8,138.0,5.0,...,True,False,True,False,True,False,True,False,False,True
7,901528,0.0,0.0,73.0,80.0,118.0,44.0,0.7,137.0,3.5,...,True,False,True,False,True,False,True,False,False,True
8,433613,0.0,0.0,50.0,80.0,97.0,40.0,0.6,150.0,4.5,...,True,False,True,False,True,False,True,False,False,True
9,621332,1.0,0.0,59.0,80.0,303.0,35.0,1.3,122.0,3.5,...,True,False,False,True,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
382,996076,0.0,0.0,57.0,60.0,105.0,49.0,1.2,150.0,4.7,...,True,False,True,False,True,False,True,False,False,True
384,397388,2.0,2.0,63.0,100.0,280.0,35.0,3.2,143.0,3.5,...,False,True,True,False,True,False,True,False,True,False
385,789906,0.0,0.0,69.0,70.0,83.0,42.0,1.2,139.0,3.7,...,True,False,True,False,True,False,True,False,False,True
386,490253,0.0,0.0,43.0,80.0,81.0,46.0,0.6,135.0,4.9,...,True,False,True,False,True,False,True,False,False,True


In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [10]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

def drop_outliers_std(catDummies, column):
    mean = catDummies[column].mean()
    std_dev = catDummies[column].std()
    lower_bound = mean - 3 * std_dev
    upper_bound = mean + 3 * std_dev
    return catDummies[(catDummies[column] >= lower_bound) & (catDummies[column] <= upper_bound)]
for column in numerical_columns:
    catDummies = pd.merge(catDummies, drop_outliers_std(catDummies, column), how='inner')
# Print the dataset
catDummies

,unique_id,al,su,age,bp,bgr,bu,sc,sod,pot,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,881763,0.0,0.0,42.0,70.0,75.0,31.0,1.2,141.0,3.5,...,True,False,True,False,True,False,True,False,False,True
1,114717,0.0,0.0,41.0,80.0,122.0,25.0,0.8,138.0,5.0,...,True,False,True,False,True,False,True,False,False,True
2,901528,0.0,0.0,73.0,80.0,118.0,44.0,0.7,137.0,3.5,...,True,False,True,False,True,False,True,False,False,True
3,433613,0.0,0.0,50.0,80.0,97.0,40.0,0.6,150.0,4.5,...,True,False,True,False,True,False,True,False,False,True
4,621332,1.0,0.0,59.0,80.0,303.0,35.0,1.3,122.0,3.5,...,True,False,False,True,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,996076,0.0,0.0,57.0,60.0,105.0,49.0,1.2,150.0,4.7,...,True,False,True,False,True,False,True,False,False,True
129,397388,2.0,2.0,63.0,100.0,280.0,35.0,3.2,143.0,3.5,...,False,True,True,False,True,False,True,False,True,False
130,789906,0.0,0.0,69.0,70.0,83.0,42.0,1.2,139.0,3.7,...,True,False,True,False,True,False,True,False,False,True
131,490253,0.0,0.0,43.0,80.0,81.0,46.0,0.6,135.0,4.9,...,True,False,True,False,True,False,True,False,False,True


## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [16]:
# Normalize the all Numerical Attributes in the dataset.
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']
#I wish there was more guidence for appropiate ways to normalize given the data, as someone who has zero experience with this
numeric = catDummies.select_dtypes(include=[int, float])
minmax = (numeric - numeric.min()) / (numeric.max() - numeric.min())
normalized = catDummies.copy()
normalized[numeric.columns] = minmax
# Print the dataset
normalized

Index(['unique_id', 'al', 'su', 'age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot',
       'hemo', 'pcv', 'wbcc', 'rbcc', 'rbc_abnormal', 'rbc_normal',
       'pc_abnormal', 'pc_normal', 'pcc_notpresent', 'pcc_present',
       'ba_notpresent', 'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes',
       'cad_no', 'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes',
       'ane_no', 'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer:

In [17]:
#Remove that column
#I am going to hope that it is appropiate to remove the ID because chat said it was ok
finalSet = normalized.drop(columns=['unique_id'])
# Print the dataset
finalSet

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.00,0.0,0.486486,0.25,0.021459,0.137255,0.098765,0.700000,0.206897,0.871287,...,True,False,True,False,True,False,True,False,False,True
1,0.00,0.0,0.472973,0.50,0.223176,0.098039,0.049383,0.600000,0.724138,0.930693,...,True,False,True,False,True,False,True,False,False,True
2,0.00,0.0,0.905405,0.50,0.206009,0.222222,0.037037,0.566667,0.206897,0.702970,...,True,False,True,False,True,False,True,False,False,True
3,0.00,0.0,0.594595,0.50,0.115880,0.196078,0.024691,1.000000,0.551724,0.643564,...,True,False,True,False,True,False,True,False,False,True
4,0.25,0.0,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.206897,0.267327,...,True,False,False,True,True,False,True,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
128,0.00,0.0,0.689189,0.00,0.150215,0.254902,0.098765,1.000000,0.620690,0.792079,...,True,False,True,False,True,False,True,False,False,True
129,0.50,0.5,0.770270,1.00,0.901288,0.163399,0.345679,0.766667,0.206897,0.524752,...,False,True,True,False,True,False,True,False,True,False
130,0.00,0.0,0.851351,0.25,0.055794,0.209150,0.098765,0.633333,0.275862,0.841584,...,True,False,True,False,True,False,True,False,False,True
131,0.00,0.0,0.500000,0.50,0.047210,0.235294,0.024691,0.500000,0.689655,0.613861,...,True,False,True,False,True,False,True,False,False,True


## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [18]:
# Export the dataframe to a new csv file
finalSet.to_csv('cleanData.csv', index=False)
# Export the dataframe to a new json file
finalSet.to_json('cleanData.json', orient='records', lines=True)

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

In [19]:
finalSet.iloc[:15]

,al,su,age,bp,bgr,bu,sc,sod,pot,hemo,...,cad_no,cad_yes,appet_good,appet_poor,pe_no,pe_yes,ane_no,ane_yes,Target_ckd,Target_notckd
0,0.00,0.0,0.486486,0.25,0.021459,0.137255,0.098765,0.700000,0.206897,0.871287,...,True,False,True,False,True,False,True,False,False,True
1,0.00,0.0,0.472973,0.50,0.223176,0.098039,0.049383,0.600000,0.724138,0.930693,...,True,False,True,False,True,False,True,False,False,True
2,0.00,0.0,0.905405,0.50,0.206009,0.222222,0.037037,0.566667,0.206897,0.702970,...,True,False,True,False,True,False,True,False,False,True
3,0.00,0.0,0.594595,0.50,0.115880,0.196078,0.024691,1.000000,0.551724,0.643564,...,True,False,True,False,True,False,True,False,False,True
4,0.25,0.0,0.716216,0.50,1.000000,0.163399,0.111111,0.066667,0.206897,0.267327,...,True,False,False,True,True,False,True,False,True,False
5,0.00,0.0,0.554054,0.00,0.167382,0.098039,0.086420,0.700000,0.620690,0.801980,...,True,False,True,False,True,False,True,False,False,True
6,0.00,0.0,0.621622,0.50,0.236052,0.078431,0.098765,0.633333,0.586207,0.871287,...,True,False,True,False,True,False,True,False,False,True
7,0.00,0.0,0.500000,0.50,0.257511,0.130719,0.086420,0.766667,0.724138,0.811881,...,True,False,True,False,True,False,True,False,False,True
8,0.50,0.0,0.729730,0.75,0.150215,0.281046,0.234568,0.533333,0.793103,0.336634,...,True,False,True,False,True,False,True,False,True,False
9,0.00,0.0,0.567568,0.00,0.180258,0.222222,0.098765,0.733333,0.689655,0.673267,...,True,False,True,False,True,False,True,False,False,True


### Paste the markdown table here
|   | al   | su   | age     | bp   | bgr     | bu       | sc   | sod      | pot      | hemo     | ... | cad_no | cad_yes | appet_good | appet_poor | pe_no | pe_yes | ane_no | ane_yes | Target_ckd | Target_notckd |
|---|------|------|---------|------|---------|----------|------|----------|----------|----------|-----|--------|---------|------------|------------|-------|--------|--------|---------|-------------|----------------|
| 0 | 0.00 | 0.0  | 0.486486| 0.25 | 0.021459| 0.137255 | 0.098765| 0.700000 | 0.206897 | 0.871287 | ... | True   | False   | True       | False      | True  | False  | True   | False   | False       | True          |
| 1 | 0.00 | 0.0  | 0.472973| 0.50 | 0.223176| 0.098039 | 0.049383| 0.600000 | 0.724138 | 0.930693 | ... | True   | False   | True       | False      | True  | False  | True   | False   | False       | True          |
| 2 | 0.00 | 0.0  | 0.905405| 0.50 | 0.206009| 0.222222 | 0.037037| 0.566667 | 0.206897 | 0.702970 | ... | True   | False   | True       | False      | True  | False  | True   | False   | False       | True          |
| 3 | 0.00 | 0.0  | 0.594595| 0.50 | 0.115880| 0.196078 | 0.024691| 1.000000 | 0.551724 | 0.643564 | ... | True   | False   | True       | False      | True  | False  | True   | False   | False       | True          |
| 4 | 0.25 | 0.0  | 0.716216| 0.50 | 1.000000| 0.163399 | 0.111111| 0.066667 | 0.206897 | 0.267327 | ... | True   | False   | False      | True       | True  | False  | True   | False   | True        | False         |
| 5 | 0.00 | 0.0  | 0.554054| 0.00 | 0.167382| 0.098039 | 0.086420| 0.700000 | 0.620690 | 0.801980 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 6 | 0.00 | 0.0  | 0.621622| 0.50 | 0.236052| 0.078431 | 0.098765| 0.633333 | 0.586207 | 0.871287 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 7 | 0.00 | 0.0  | 0.500000| 0.50 | 0.257511| 0.130719 | 0.086420| 0.766667 | 0.724138 | 0.811881 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 8 | 0.50 | 0.0  | 0.729730| 0.75 | 0.150215| 0.281046 | 0.234568| 0.533333 | 0.793103 | 0.336634 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 9 | 0.00 | 0.0  | 0.567568| 0.00 | 0.180258| 0.222222 | 0.098765| 0.733333 | 0.689655 | 0.673267 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 10| 0.00 | 0.0  | 0.864865| 0.50 | 0.017167| 0.202614 | 0.012346| 0.766667 | 0.551724 | 0.732673 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 11| 0.00 | 0.0  | 0.878378| 0.25 | 0.047210| 0.052288 | 0.049383| 0.833333 | 0.724138 | 0.693069 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 12| 0.00 | 0.0  | 0.662162| 0.50 | 0.270386| 0.045752 | 0.098765| 0.500000 | 0.655172 | 0.544554 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 13| 0.00 | 0.0  | 0.229730| 0.50 | 0.124464| 0.235294 | 0.098765| 0.733333 | 0.379310 | 0.990099 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |
| 14| 0.00 | 0.0  | 0.986486| 0.50 | 0.175966| 0.222222 | 0.098765| 0.866667 | 0.241379 | 0.851485 | ... | True   | False   | True       | False      | True  | False  | True   | False   | True        | False         |


** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

I guess somehow cross reference the output from the LLM with the original dataset. Maybe traverse the output of the LLM and check if it actually exists in the orignal set. This would probably be expensive. Also be very clear and explicit with how prompts are being phrased so as to get the most accurate input. 

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [20]:
# Code to rename all the columns in the dataset
# Define the renaming dictionary
rename_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}
# Rename the DataFrame columns
finalSet.rename(columns=rename_dict, inplace=True)

finalSet.columns


Index(['albumin', 'sugar', 'age', 'blood pressure', 'blood glucose random',
       'blood urea', 'serum creatinine', 'sodium', 'potassium', 'hemoglobin',
       'packed cell volume', 'wbcc', 'rbcc', 'rbc_abnormal', 'rbc_normal',
       'pc_abnormal', 'pc_normal', 'pcc_notpresent', 'pcc_present',
       'ba_notpresent', 'ba_present', 'htn_no', 'htn_yes', 'dm_no', 'dm_yes',
       'cad_no', 'cad_yes', 'appet_good', 'appet_poor', 'pe_no', 'pe_yes',
       'ane_no', 'ane_yes', 'Target_ckd', 'Target_notckd'],
      dtype='object')

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [24]:
# Converted SQL to Pandas code
# Sample DataFrame creation for demonstration
data = {
    'Target': ['ckd', 'notckd', 'ckd', 'notckd', 'ckd', 'notckd'],
    'Value': [1, 1, 1, 1, 1, 1]
}
df = pd.DataFrame(data)
# Group by 'Target' and count the occurrences
result = df.groupby('Target').size().reset_index(name='count')
result

,Target,count
0,ckd,3
1,notckd,3


*Describe what the above code does here*

The code creates a dataframe which organizes the number of occurences of each value of a target variable. The provided Pandas code groups the DataFrame by the `Target` column and counts the number of occurrences of each unique value in that column. It returns a new DataFrame that includes each unique `Target` value along with its corresponding count, creating a summary of the data distribution within the `Target` column.